chain/processor_test.go

```diff
@@ -0,0 +1,435 @@
+			workers:        workers.NewSerial(),
+			db: func() merkledb.MerkleDB {
+				db := createDB()
+				require.NoError(t, db.Put([]byte(heightKey), binary.BigEndian.AppendUint64(nil, 0)))
+				require.NoError(t, db.Put([]byte(timestampKey), []byte{}))
+				return db
+			}(),
+			createBlock: createValidBlock,
+			expectedErr: chain.ErrFailedToParseParentTimestamp,
+		},
+		{
+			name:           "non-empty block timestamp less than parent timestamp with gap",
+			validityWindow: &validitywindowtest.MockTimeValidityWindow[*chain.Transaction]{},
+			workers:        workers.NewSerial(),
+			db: func() merkledb.MerkleDB {
+				db := createDB()
+				require.NoError(t, db.Put([]byte(heightKey), binary.BigEndian.AppendUint64(nil, 0)))
+				require.NoError(t, db.Put([]byte(timestampKey), binary.BigEndian.AppendUint64(nil, 0)))
+				return db
+			}(),
+			createBlock: func(parentRoot ids.ID) *chain.StatelessBlock {
+				block, err := chain.NewStatelessBlock(
+					ids.Empty,
+					0,
+					1,
+					[]*chain.Transaction{
+						func() *chain.Transaction {
+							tx, err := chain.NewTransaction(
+								&chain.Base{},
+								[]chain.Action{},
+								&mockAuth{
+									typeID: 1,
+								},
+							)
+							require.NoError(t, err)
+							return tx
+						}(),
+					},
+					parentRoot,
+				)
+				require.NoError(t, err)
+				return block
+			},
+			expectedErr: chain.ErrTimestampTooEarly,
+		},
+		{
+			name:           "empty block timestamp less than parent timestamp with gap",
+			validityWindow: &validitywindowtest.MockTimeValidityWindow[*chain.Transaction]{},
+			workers:        workers.NewSerial(),
+			db: func() merkledb.MerkleDB {
+				db := createDB()
+				require.NoError(t, db.Put([]byte(heightKey), binary.BigEndian.AppendUint64(nil, 0)))
+				require.NoError(t, db.Put([]byte(timestampKey), binary.BigEndian.AppendUint64(nil, 0)))
+				return db
+			}(),
+			createBlock: func(parentRoot ids.ID) *chain.StatelessBlock {
+				block, err := chain.NewStatelessBlock(
+					ids.Empty,
+					0,
+					1,
+					nil,
+					parentRoot,
+				)
+				require.NoError(t, err)
+				return block
+			},
+			expectedErr: chain.ErrTimestampTooEarly,
+		},
+		{
+			name:           "failed to get fee",
+			validityWindow: &validitywindowtest.MockTimeValidityWindow[*chain.Transaction]{},
+			workers:        workers.NewSerial(),
+			db: func() merkledb.MerkleDB {
+				db := createDB()
+				require.NoError(t, db.Put([]byte(heightKey), binary.BigEndian.AppendUint64(nil, 0)))
+				require.NoError(t, db.Put([]byte(timestampKey), binary.BigEndian.AppendUint64(nil, 0)))
+				return db
+			}(),
+			createBlock: createValidBlock,
+			expectedErr: chain.ErrFailedToFetchParentFee,
+		},
+		{
+			name: "fails replay protection",
+			validityWindow: &validitywindowtest.MockTimeValidityWindow[*chain.Transaction]{
+				OnVerifyExpiryReplayProtection: func(_ context.Context, _ validitywindow.ExecutionBlock[*chain.Transaction]) error {
```

Comment:
nit: delete unnamed param names, not sure why our linter doesn't catch this
```suggestion
				OnVerifyExpiryReplayProtection: func(context.Context, validitywindow.ExecutionBlock[*chain.Transaction]) error {
```

Reply:
Done